### PR TITLE
Port NumPy typing testing style to PyTorch

### PR DESCRIPTION
**Pull request by guilhermeleobas**
This is a follow-up PR of #52408 and includes the `pass/` and `fail/` directories.

### CODE REVIEW

**Comment by guilhermeleobas** [comment]
Hi @rgommers, can you do an initial review? I'll add more examples to the `fail/` dir in the upcoming days.

File: test/test_typing.py [hunk 1]
```diff
@@ -35,6 +38,12 @@ def _key_func(key: str) -> str:
     return os.path.join(drive, tail.split(":", 1)[0])
 
 
+def _strip_filename(msg: str) -> str:
+    """Strip the filename from a mypy message."""
```
Comment (rgommers) [inline test/test_typing.py:42]: This change is helpful

File: test/typing/pass/creation_ops.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+# From the docs, there are quite a few ways to create a tensor:
+# https://pytorch.org/docs/stable/tensors.html
+
+# torch.tensor()
+torch.tensor([[0.1, 1.2], [2.2, 3.1], [4.9, 5.2]])
+torch.tensor([0, 1])
+torch.tensor([[0.11111, 0.222222, 0.3333333]],
+                         dtype=torch.float64,
+                         device=torch.device('cuda:0'))
```
Comment (rgommers) [inline test/typing/pass/creation_ops.py:15]: minor: funky indentation here. also further down in multiple places. maybe you had an editor setup issue? would be good to clean up

File: test/typing/pass/creation_ops.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+x = torch.randn(3, 3)
+torch.as_strided(x, (2, 2), (1, 2))
+torch.as_strided(x, (2, 2), (1, 2), 1)
+
+# torch.from_numpy
+if TEST_NUMPY:
+    torch.from_numpy(np.array([1, 2, 3]))
+
+# torch.zeros/zeros_like
+torch.zeros(2, 3)
```
Comment (rgommers) [inline test/typing/pass/creation_ops.py:46]: would be good to also add
```
torch.zeros((2, 3))
torch.zeros([2, 3])
```
to ensure list and tuple input work as advertised.

File: test/typing/pass/math_ops.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+import torch
+import math
+
+a = torch.randn(4)
+b = torch.randn(4)
+t = torch.tensor([-1, -2, 3], dtype=torch.int8)
+
+# abs/absolute
+torch.abs(torch.Tensor([-1, -2, 3]))
+torch.absolute(torch.Tensor([-1, -2, 3]))
```
Comment (rgommers) [inline test/typing/pass/math_ops.py:11]: Please don't use `torch.Tensor`, it's the deprecated constructor and we just got rid of it everywhere. use `torch.tensor` instead

File: test/typing/pass/math_ops.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+
+# log2
+torch.log2(torch.rand(5))
+
+# logaddexp
+torch.logaddexp(torch.tensor([-1.0]), torch.tensor([-1.0, -2, -3]))
+torch.logaddexp(torch.tensor([-100.0, -200, -300]), torch.tensor([-1.0, -2, -3]))
+torch.logaddexp(torch.tensor([1.0, 2000, 30000]), torch.tensor([-1.0, -2, -3]))
+
+# logaddexp2
```
Comment (rgommers) [inline test/typing/pass/math_ops.py:184]: missing code here?

File: test/typing/pass/math_ops.py [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+torch.nan_to_num(x)
+torch.nan_to_num(x, nan=2.0)
+torch.nan_to_num(x, nan=2.0, posinf=1.0)
+
+# neg/negative
+torch.neg(torch.randn(5))
+
+# nextafter
+eps = torch.finfo(torch.float32).eps
+torch.nextafter(torch.Tensor([1, 2]), torch.Tensor([2, 1])) == torch.Tensor([eps + 1, 2 - eps])
```
Comment (rgommers) [inline test/typing/pass/math_ops.py:248]: more `Tensor` -> `tensor`

File: test/typing/reveal/tensor_sampling.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
+# flake8: noqa
+import torch
+
+# seed
+# manual_seed
+# initial_seed
+# get_rng_state
+# set_rng_state
```
Comment (rgommers) [inline test/typing/reveal/tensor_sampling.py:8]: Were you planning to finish these, or are they left like this on purpose?

Comment (rgommers) [reply test/typing/reveal/tensor_sampling.py:8]: There are some more further down.

File: torch/quasirandom.py [hunk 8]
```diff
@@ -102,7 +103,8 @@ def draw(self, n=1, out=None, dtype=torch.float32):
 
         return result
 
-    def draw_base2(self, m, out=None, dtype=torch.float32):
+    def draw_base2(self, m: int, out: Optional[torch.Tensor] = None,
+                   dtype: torch.dtype = torch.float32):
```
Comment (rgommers) [inline torch/quasirandom.py:107]: missing return type

**Review by rgommers — CHANGES_REQUESTED** [review]
Thanks @guilhermeleobas, overall this looks good. A number of small things to fix. Would also be good to merge with master to get a fresh CI run.

**Comment by guilhermeleobas** [comment]
Hi Ralf, I've addressed all your comments

**Review by rgommers — APPROVED** [review]
Thanks @guilhermeleobas, this LGTM now.

**Review by malfet — APPROVED** [review]

**Review by walterddr — APPROVED** [review]

**Pull request closed without merging**